processing/src/main/java/io/druid/segment/DimensionIndexer.java

```diff
@@ -0,0 +1,350 @@
+import io.druid.query.dimension.DimensionSpec;
+import io.druid.query.filter.DruidPredicateFactory;
+import io.druid.query.filter.ValueMatcher;
+import io.druid.segment.data.Indexed;
+import io.druid.segment.incremental.IncrementalIndex;
+import io.druid.segment.incremental.IncrementalIndexStorageAdapter;
+
+/**
+ * Processing related interface
+ *
+ * A DimensionIndexer is a per-dimension stateful object that encapsulates type-specific operations and data structures
+ * used during the in-memory ingestion process (i.e., work done by IncrementalIndex).
+ *
+ * Ingested row values are passed to a DimensionIndexer, which will update its internal data structures such as
+ * a value->ID dictionary as row values are seen.
+ *
+ * The DimensionIndexer is also responsible for implementing various value lookup operations,
+ * such as conversion between an encoded value and its full representation. It maintains knowledge of the
+ * mappings between encoded values and actual values.
+ *
+ *
+ * Sorting and Ordering
+ * --------------------
+ * When encoding is present, there are two relevant orderings for the encoded values.
+ *
+ * 1.) Ordering based on encoded value's order of ingestion
+ * 2.) Ordering based on converted actual value
+ *
+ * Suppose we have a new String dimension DimA, which sees the values "Hello", "World", and "Apple", in that order.
+ * This would correspond to dictionary encodings of "Hello"=0, "World"=1, and "Apple"=2, by the order
+ * in which these values were first seen during ingestion.
+ *
+ * However, some use cases require the encodings to be sorted by their associated actual values.
+ * In this example, that ordering would be "Apple"=0, "Hello"=1, "World"=2.
+ *
+ * The first ordering will be referred to as "Unsorted" in the documentation for this interface, and
+ * the second ordering will be referred to as "Sorted".
+ *
+ * The unsorted ordering is used during ingestion, within the IncrementalIndex's TimeAndDims keys; the encodings
+ * are built as rows are ingested, taking the order in which new dimension values are seen.
+ *
+ * The generation of a sorted encoding takes place during segment creation when indexes are merged/persisted.
+ * The sorted ordering will be used for dimension value arrays in that context and when reading from
+ * persisted segments.
+ *
+ * Note that after calling the methods below that deal with sorted encodings,
+ * - getSortedEncodedValueFromUnsorted()
+ * - getUnsortedEncodedValueFromSorted()
+ * - getSortedIndexedValues()
+ * - convertUnsortedEncodedArrayToSortedEncodedArray()
+ *
+ * calling processRowValsToUnsortedEncodedArray() afterwards can invalidate previously read sorted encoding values
+ * (i.e., new values could be added that are inserted between existing values in the ordering).
+ *
+ *
+ * Thread Safety
+ * --------------------
+ * Each DimensionIndexer exists within the context of a single IncrementalIndex. Before IndexMerger.persist() is
```

Comment:
I think the intent is not that it becomes non-thread-safe after persist is called, but rather:

- it is expected that a single persist thread is the only thing calling the methods that deal with getting sorted value; i.e. multithreaded persisting is not ok
- after that thread starts calling those methods, writes are no longer safe